File: python/src/data_structure/linked_list.py
"""Here we will create a Linked list"""

import json
from typing import Generic, List, Literal, Optional, TypeVar

from ..model.node import Node

T = TypeVar("T")


class LinkedList(Generic[T]):
    """Defines a Linked List"""

    def __init__(self) -> None:
        """
        Creates an instance of `Linked List`
        :type value: None
        """
        self._head: Optional[Node[T]] = None
        self._size = 0

    def push(self, data: T) -> None:
        """
        Add a node at the end of the `Linked List`
        :param data: T
        :type data: T
        """

        new_node = Node(data)
        if not self._head:
            self._head = new_node
            self._size += 1
            return

        current = self._head
        while current.right:
            current = current.right
        current.right = new_node
        self._size += 1

    def un_shift(self, data: T) -> int:
        """
        Add a node at the start of the `Linked List`
        :param data: T
        :type data: T
        :return: The size of the `Linked List`
        """

        new_node = Node(data)
        if not self._head:
            self._head = new_node
            self._size += 1
            return self._size

        new_node.right = self._head
        self._head = new_node
        self._size += 1
        return self._size

    def pop(self) -> Optional[T]:
        """
        Remove the last node of the `Linked List`.
        :return: The last value in the list
        """
        if not self._head:
            return None

        current = self._head
        while current.right:
            if not current.right.right:
                value = current.right.value
                current.right = None
                self._size -= 1
                return value

            current = current.right

        self._head = None
        self._size -= 1
        return current.value

    def shift(self) -> Optional[T]:
        """
        Remove the first node of the `Linked List`.
        :return: The first value in the list
        """
        if not self._head:
            return None

        value = self._head.value
        self._head = self._head.right
        self._size -= 1

        return value

    def insert_at(self, data: T, position: int) -> None:
        """
        It inserts data at a given index.

        :param data: The data to insert into the list
        :type data: T
        :param position: the position of the data to insert
        :type position: int
        :return: None
        """

        if not self._size >= position >= 0:
            return

        new_node = Node(data)

        if not self._head or position == 0:
            if position == 0:
                new_node.right = self._head
            self._head = new_node
            self._size += 1
            return

        current = self._head
        count = 0
        while count < position - 1 and current.right:
            count += 1
            current = current.right

        new_node.right = current.right
        current.right = new_node
        self._size += 1

    def print_all(self, type_as: Literal["nested", "arrow"] = "arrow") -> None:
        """
        This will print all the items in the `Linked List`.
        """
        if not self._head:
            print("No item found!")
            return

        if type_as == "nested":
            print(json.dumps(self._head.to_object(), indent=2))
            return

        current = self._head
        items: List[T] = []
        items.append(current.value)

        while current.right:
            current = current.right
            items.append(current.value)

        print("List: ", " -> ".join(map(str, items)))
        return

    @property
    def size(self) -> int:
        """
        The is the `getter` for the `size`.
        :return: The value of the instance variable `_size`.
        """
        return self._size

    def __repr__(self) -> str:
        """
        Official representation of class

        The __repr__ function returns a string representation of the object
        :return: The class name.
        """
        return "<class 'LinkedList'>"

    def __str__(self) -> str:
        """
        The magic method provides string representation of instance

        The `__str__` function is a special function that is called when you try to print an object
        """
        return f"Value: {self}"

File: typescript/src/data-structure/linked-list.ts
import { MyNode } from "model";
import { NotFunction } from "types";

export default class LinkedList<T> {
  #head?: MyNode<T> = undefined;
  #size = 0;

  /**
   * Add a node at the end of the `Linked List`.
   * @param data - NotFunction<T> - This is the node that we want to add to the end of the list.
   * @returns The size of the list
   */
  push(data: NotFunction<T>) {
    const newNode = new MyNode(data);

    if (!this.#head) {
      this.#head = newNode;
      return ++this.#size;
    }

    let current = this.#head;

    while (current.right) {
      current = current.right;
    }

    current.right = newNode;
    return ++this.#size;
  }

  /**
   * Add a node at the start of the `Linked List`.
   * @param data - NotFunction<T>
   * @returns The size of the list
   */
  unshift(data: NotFunction<T>) {
    const newNode = new MyNode(data);

    if (!this.#head) {
      this.#head = newNode;
      return ++this.#size;
    }

    newNode.right = this.#head;
    this.#head = newNode;
    return ++this.#size;
  }

  /**
   * Remove the last node of the `Linked List`.
   * @returns The last value in the list.
   */
  pop(): T | undefined {
    if (!this.#head) {
      return undefined;
    }

    let current = this.#head;

    while (current.right) {
      if (!current.right.right) {
        const value = current.right.value;
        current.right = null;
        this.#size--;
        return value;
      }

      current = current.right;
    }
    this.#head = undefined;
    this.#size--;
    return current.value;
  }

  /**
   * Remove the first node of the `Linked List`.
   * @returns The first value in the list
   */
  shift(): T | undefined {
    if (!this.#head) {
      return undefined;
    }

    const value = this.#head.value;
    this.#head = this.#head.right || undefined;
    this.#size--;
    return value;
  }

  /**
   * It inserts data at a given position.
   * @param data - The data to be inserted.
   * @param {number} position - The position in the `LinkedList` to insert the data.
   */
  insertAt(data: NotFunction<T>, position: number) {
    if (position > this.#size || position < 0) {
      return;
    }

    const newNode = new MyNode(data);

    if (!this.#head || position === 0) {
      if (position === 0) {
        newNode.right = this.#head || null;
      }

      this.#head = newNode;
      this.#size++;
      return;
    }

    let current = this.#head;
    let count = 0;
    while (count < position - 1 && current.right) {
      count += 1;
      current = current.right;
    }

    newNode.right = current.right;
    current.right = newNode;
    this.#size++;
  }

  /**
   * Print all the items of the `LinkedList`.
   * @param {"nested" | "arrow"} [type=arrow] - "nested" | "arrow" = "arrow"
   * @returns The values of the linked list.
   */
  printAll(type: "nested" | "arrow" = "arrow") {
    if (!this.#head) {
      console.log("The linked list is empty");
      return;
    }

    if (type === "nested") {
      console.log(JSON.stringify(this.#head.toObject(), null, 2));
      return;
    }

    const list: T[] = [];
    let current = this.#head;
    list.push(current.value);

    while (current.right) {
      current = current.right;
      list.push(current.value);
    }
    console.log("List: ", list.join(" -> "));
  }

  /**
   * The size property is a getter that returns the value of the private `#size` property
   * @returns The size of the `LinkedList`.
   */
  get size() {
    return this.#size;
  }
}
